addons/mail/models/mail_activity.py

```diff
@@ -572,53 +585,161 @@ def action_open_document(self):
-        if domain or limit or offset:
-            res = self.env[res_model].search(domain or [], limit=limit, offset=offset)
-            activity_domain.append(('res_id', 'in', res.ids))
-        grouped_activities = self.env['mail.activity']._read_group(
+        is_filtered = domain or limit or offset
+        domain_res_ids = Model._search(domain or [], offset, limit) if is_filtered else []
+        if is_filtered:
+            activity_domain.append(('res_id', 'in', domain_res_ids))
+        all_activities = self.env['mail.activity'].with_context(active_test=not fetch_done).search_read(
             activity_domain,
-            ['res_id', 'activity_type_id'],
-            ['id:array_agg', 'date_deadline:min', '__count'])
-        # filter out unreadable records
-        if not domain:
-            res_ids = tuple(res_id for res_id, *_ in grouped_activities)
-            res_ids_set = set(self.env[res_model].search([('id', 'in', res_ids)])._ids)
-            grouped_activities = [a for a in grouped_activities if a[0] in res_ids_set]
+            ['activity_type_id', 'res_id', 'date_deadline', 'user_id', 'active'] +
+            (['attachment_ids'] if fetch_done else []),
+        )
+        all_ongoing_activities, all_completed_activities = partition(lambda a: a['active'], all_activities)
+
+        # 2. Get attachment of completed activities
+        if all_completed_activities:
+            all_attachment_ids = [
+                attachment_id
+                for act in all_completed_activities
+                for attachment_id in act['attachment_ids']
+            ]
+            attachments_by_id = {
+                a['id']: a
+                for a in self.env['ir.attachment'].search_read([['id', 'in', all_attachment_ids]],
+                                                               ['create_date', 'name'])
+            } if all_attachment_ids else {}
+        else:
+            attachments_by_id = {}
+
+        # 3. Group activities per records and activity type
+        grouped_completed_activities = dict(groupby(all_completed_activities,
+                                                    key=lambda a: (a['res_id'], a['activity_type_id'][0])))
+        grouped_ongoing_activities = dict(groupby(all_ongoing_activities,
+                                                  key=lambda a: (a['res_id'], a['activity_type_id'][0])))
+
+        # 4. Filter out unreadable records
+        all_res_id_activity_type_id = grouped_ongoing_activities.keys() | grouped_completed_activities.keys()
+        if not is_filtered:
+            res_ids = list({res_id for res_id, __ in all_res_id_activity_type_id})
+            res_ids_set = set(Model.search([('id', 'in', res_ids)])._ids)
+            all_res_id_activity_type_id = [(res_id, activity_type_id)
+                                           for res_id, activity_type_id in all_res_id_activity_type_id
+                                           if res_id in res_ids_set]
```

Comment:
Try to keep a simple naming and coding style when possible, as you go through a lot of intermediate variables and type convert that can make code a bit hard to follow, like
```suggestion
        res_ids_types_tuples = grouped_ongoing.keys() | grouped_completed.keys()
        if not is_filtered:
            filtered = Model.search([('id', 'in', [r[0] for r in res_ids_types_tuples)]).ids
            res_ids_types_tuples = list(filter(lambda x: x[0] in filtered, res_ids_types_tuples))
``